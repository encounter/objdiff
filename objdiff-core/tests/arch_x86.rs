use objdiff_core::{diff, diff::display::SymbolFilter, obj};

mod common;

#[test]
#[cfg(feature = "x86")]
fn read_x86() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/staticdebug.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?PrintThing@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_combine_sections() {
    let diff_config = diff::DiffObjConfig {
        combine_data_sections: true,
        combine_text_sections: true,
        ..Default::default()
    };
    let obj = obj::read::parse(include_object!("data/x86/rtest.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj.sections);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_64() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86_64/vs2022.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "?Dot@Vector@@QEAAMPEAU1@@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn display_section_ordering() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/basenode.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    let obj_diff =
        diff::diff_objs(Some(&obj), None, None, &diff_config, &diff::MappingConfig::default())
            .unwrap()
            .left
            .unwrap();
    let section_display =
        diff::display::display_sections(&obj, &obj_diff, SymbolFilter::None, false, false, false);
    insta::assert_debug_snapshot!(section_display);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_jumptable() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/jumptable.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?test@@YAHH@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

// Inferred size of functions should ignore symbols with specific prefixes
#[test]
#[cfg(feature = "x86")]
fn read_x86_local_labels() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/local_labels.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
}
