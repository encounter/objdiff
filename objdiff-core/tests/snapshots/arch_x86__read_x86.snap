---
source: objdiff-core/tests/arch_x86.rs
expression: obj
---
Object {
    arch: ArchX86 {
        arch: X86,
        endianness: Little,
    },
    endianness: Little,
    symbols: [
        Symbol {
            name: "objdiffstaticdebug.cpp",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "@comp.id",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Object,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.drectve]",
            demangled_name: None,
            address: 0,
            size: 38,
            kind: Section,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.data]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                1,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "?a@@3PAXA",
            demangled_name: Some(
                "void *a",
            ),
            address: 0,
            size: 4,
            kind: Object,
            section: Some(
                1,
            ),
            flags: FlagSet(Global | SizeInferred),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "[.text]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "?PrintThing@@YAXXZ",
            demangled_name: Some(
                "void __cdecl PrintThing(void)",
            ),
            address: 0,
            size: 18,
            kind: Function,
            section: Some(
                2,
            ),
            flags: FlagSet(Local | SizeInferred),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "_printf",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Function,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: None,
        },
        Symbol {
            name: "$SG526",
            demangled_name: None,
            address: 4,
            size: 6,
            kind: Object,
            section: Some(
                1,
            ),
            flags: FlagSet(Local | SizeInferred),
            align: None,
            virtual_address: None,
        },
    ],
    sections: [
        Section {
            id: ".drectve-0",
            name: ".drectve",
            address: 0,
            size: 38,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            align: Some(
                1,
            ),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".data-0",
            name: ".data",
            address: 0,
            size: 10,
            kind: Data,
            data: SectionData(
                10,
            ),
            flags: FlagSet(),
            align: Some(
                4,
            ),
            relocations: [
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 0,
                    target_symbol: 6,
                    addend: 0,
                },
            ],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".text-0",
            name: ".text",
            address: 0,
            size: 18,
            kind: Code,
            data: SectionData(
                18,
            ),
            flags: FlagSet(),
            align: Some(
                16,
            ),
            relocations: [
                Relocation {
                    flags: Coff(
                        6,
                    ),
                    address: 4,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Coff(
                        20,
                    ),
                    address: 9,
                    target_symbol: 7,
                    addend: 0,
                },
            ],
            line_info: {},
            virtual_address: None,
        },
    ],
    split_meta: None,
    path: None,
    timestamp: None,
    flow_analysis_results: {},
}
